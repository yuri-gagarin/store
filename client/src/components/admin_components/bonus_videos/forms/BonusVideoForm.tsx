import React, { useState, useEffect, useRef }  from "react";
import { Button, Form, TextArea } from "semantic-ui-react";

export type FormState = {
  description: string;
  youTubeURL: string;
  vimeoURL: string;
}

interface Props {
  description: string;
  youTubeURL: string;
  vimeoURL: string;
  handleCreateProduct(data: FormState): void;
  handleUpdateProduct(data: FormState): void;
}


const ProductForm: React.FC<Props> = ({ description, youTubeURL, vimeoURL, handleCreateProduct, handleUpdateProduct }): JSX.Element => {
  
  const [ newForm, setNewForm ] = useState<boolean>(true)
  const [ formState, setFormState ] = useState<FormState>({ description, youTubeURL, vimeoURL });
  const bonusVideoFormRef = useRef<HTMLDivElement>(document.createElement("div"));
  
  const handleYouTubeURLChange = (e: React.ChangeEvent<HTMLInputElement>):void => {
    setFormState({
      ...formState,
      youTubeURL: e.target.value
    });
  };
  const handleVimeoURLChange = (e: React.ChangeEvent<HTMLInputElement>):void => {
    setFormState({
      ...formState,
      vimeoURL: e.target.value
    });
  };
  const hadnleDescriptionChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    setFormState({
      ...formState,
      description: e.target.value
    });
  };
  
  const handleSubmit = (e: React.MouseEvent<HTMLButtonElement>) => {
    e.preventDefault();
    if (newForm) {
      handleCreateProduct(formState);
    } else {
      handleUpdateProduct(formState);
    }
  };

  useEffect(() => {
    if (description || youTubeURL || vimeoURL) {
      setNewForm(false);
    } else {
      setNewForm(true);
    }
  },  [description, youTubeURL, vimeoURL]);

  useEffect(() => {
    if (bonusVideoFormRef.current) {
      const elem = bonusVideoFormRef.current.getBoundingClientRect();
      window.scrollTo({
        top: elem.bottom,
        behavior: "smooth"
      })
    }
  }, [bonusVideoFormRef]);
  
  return (
    <div className="createProductFormHolder" ref={bonusVideoFormRef}>
      <Form id="createProductForm">
        <Form.Field>
          <label>Video YouTube URL:</label>
          <input 
            onChange={handleYouTubeURLChange} 
            placeholder="YouTube URL here ..." 
            value={formState.youTubeURL}
          />
        </Form.Field>
        <Form.Field>
          <label>Video Vimeo URL:</label>
          <input 
            onChange={handleVimeoURLChange} 
            placeholder="Vimeo URL here..." 
            value={formState.vimeoURL}
          />
        </Form.Field>
        <Form.Field
          id='form-textarea-control-opinion'
          control={TextArea}
          label='Video Description'
          onChange={hadnleDescriptionChange}
          placeholder='Video description here...'
          value={formState.description}
         />
         {
           newForm 
            ? <Button type='submit' onClick={handleSubmit} content= "Save New Bonus Video" />
            : <Button type='submit' onClick={handleSubmit} content= "Update Bonus Video" />
         }  
      </Form>
    </div>
  );
};

export default ProductForm;