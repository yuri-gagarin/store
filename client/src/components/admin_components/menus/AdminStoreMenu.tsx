import React, { useState, useEffect } from "react";
import { Menu, MenuItemProps } from "semantic-ui-react";
// routing //
import { withRouter, RouteComponentProps } from "react-router-dom"
// css imports //
import "./css/adminStoreMenu.css";

interface Props extends RouteComponentProps {

}
const AdminStoreMenu: React.FC<Props> = ({ history }): JSX.Element => {
  const [ activeItem, setActiveItem ] = useState<string>("view_all");
  const [ menuOpen, setMenuOpen ] = useState<boolean>(false);

  const handleItemClick = (e: React.MouseEvent, { name }: MenuItemProps): void => {
    setActiveItem(String(name));
    const baseUrl: string = "/admin/home/my_store";
    switch (name) {
      case "view_all": 
        history.push(baseUrl + "/all");
        break;
      case "create":
        history.push(baseUrl + "/create");
        break;
      case "manage":
        history.push(baseUrl + "/manage");
        break;
      default: history.push(baseUrl);
    }
  }
  useEffect(() => {
    setTimeout(() => {
      setMenuOpen(true);
    }, 200);
  }, []);

  return (
    <div>
      <Menu tabular className={ menuOpen ? "adminStoreMenu storeMenuOpen" : "adminStoreMenu" }>
      <Menu.Item
          name='view_all'
          content="View All"
          active={activeItem === 'view_all'}
          onClick={handleItemClick}
        />
        <Menu.Item
          name='create'
          content="Create Store"
          active={activeItem === 'create'}
          onClick={handleItemClick}
        />
        <Menu.Item
          name='manage'
          content="Manage"
          active={activeItem === 'manage'}
          onClick={handleItemClick}
        />
      </Menu>
    </div>
  )
};

export default withRouter(AdminStoreMenu);