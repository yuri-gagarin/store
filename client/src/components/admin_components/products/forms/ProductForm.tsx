import React, { useState, useEffect, useRef, memo }  from "react";
import { Button, Form, TextArea } from "semantic-ui-react";
// type definitions//
import { ProductFormState } from "../type_definitions/productTypes";

interface Props {
  name: string;
  price: string;
  description: string;
  details: string;
  newForm: boolean;
  handleCreateProduct(data: ProductFormState): void;
  handleUpdateProduct(data: ProductFormState): void;
}
class ProductForm extends React.Component<Props, ProductFormState> {
  constructor(props: Props) {
    super(props)
    // this.productFormRef = useRef<HTMLDivElement>(document.createElement("div"));
    this.state = {
      name: "",
      price: "",
      description: "",
      details: "",
    }

  }
  // local form state //
  // form ref //
  
  // form state changes listeners //
  handleNameChange = (e: React.ChangeEvent<HTMLInputElement>):void => {
    this.setState({
      ...this.state,
      name: e.target.value
    });
  };
  handlePriceChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    this.setState({
      ...this.state,
      price: e.target.value
    });
  };
  hadnleDescriptionChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    this.setState({
      ...this.state,
      description: e.target.value
    });
  };
  handleDetailsChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    this.setState({
      ...this.state,
      details: e.target.value
    });
  };
  // submit action //
  handleSubmit = (e: React.MouseEvent<HTMLButtonElement>) => {
    e.preventDefault();
    if (this.props.newForm) {
      this.props.handleCreateProduct(this.state);
    } else {
      this.props.handleUpdateProduct(this.state);
    }
  };
  componentDidMount() {

  }
  shouldComponentUpdate(nextProps: Props, nextState: ProductFormState) {
    console.log(nextProps)
    const { name: newName } = nextProps;
    if (newName === this.props.name) {
      console.log("shouldnt");
      return false
    }
    return true;
  }

  // lifecycle hooks //
  /*
  useEffect(() => {
    if (productFormRef.current) {
      const elem = productFormRef.current.getBoundingClientRect();
      window.scrollTo({
        top: elem.bottom,
        behavior: "smooth"
      })
    }
  }, [productFormRef]);
  */
  render() {
    const { name, price, description, details } = this.state;
    const { newForm } = this.props;
    return (
      <div className="productFormDiv">
        <Form id="productForm">
          <Form.Field>
            <label>Product name</label>
            <input 
              id="adminProductFormNameInput"
              onChange={this.handleNameChange} 
              placeholder="Product name here ..." 
              value={name}
            />
          </Form.Field>
          <Form.Field>
            <label>Product price</label>
            <input 
              id="adminProductFormPriceInput"
              onChange={this.handlePriceChange} 
              placeholder="Product price here..." 
              value={price}
            />
          </Form.Field>
          <Form.Field
            id='adminProductFormDescInput'
            control={TextArea}
            label='Product Description'
            onChange={this.hadnleDescriptionChange}
            placeholder='Product description here...'
            value={description}
           />
          <Form.Field
            id='adminProductFormDetailsInput'
            control={TextArea}
            label='Product Details'
            onChange={this.handleDetailsChange}
            placeholder='Product details here...'
            value={details}
          />
        </Form>
        {
          newForm 
            ? <Button 
                id="adminProductFormCreate" 
                type="button"
                onClick={this.handleSubmit} 
                content= "Create  New Product" 
              />
            : <Button 
              id="adminProductFormUpdate" 
              type='button'
              onClick={this.handleSubmit} 
              content="Update Product" 
            />
        }  
      </div>
    );
  }
}

export default ProductForm;