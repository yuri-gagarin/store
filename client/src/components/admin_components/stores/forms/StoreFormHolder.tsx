import React, { useState, useEffect } from "react";
import { Button, Grid } from "semantic-ui-react";
// css imports //
import "./css/storeFormHolder.css";
// additional components //
import StoreForm from "./StoreForm";
import StoreImgPreviewHolder from "../image_preview/StoreImgPreviewThumbs";
import StoreImageUplForm from "./StoreImageUplForm";
// state //
import { IGlobalAppState, AppAction } from "../../../../state/Store";
// api actions //
import { createStore, editStore } from "../actions/APIstoreActions";
// helpers //
import { ConvertDate } from "../../../helpers/displayHelpers";

interface Props {
  state: IGlobalAppState;
  dispatch: React.Dispatch<AppAction>;
}

const StoreFormHolder: React.FC<Props> = ({ state, dispatch }): JSX.Element => {
  const [ formOpen, setFormOpen ] = useState<boolean>(false);
  const [ imgUpload, setImgUpload ] = useState<boolean>(false);
  const [ newForm, setNewForm ] = useState<boolean>(true);

  const { currentStoreData } = state.storeState;
  const { title, description } = currentStoreData;

  const handleCreateStore = (title: string, description: string): void => {
    const storePams = {
      title,
      description,
      storeImages: currentStoreData.images
    };
    createStore(storePams, dispatch)
      .then((success) => {
        if (success) {
          // store created //
          setImgUpload(true);
        } else {
          console.error("error");
        }
      });
  };

  const handleUpdateStore = (title: string, description: string): void => {
    const storeParams = {
      title, description, storeImages: currentStoreData.images
    }
    editStore(currentStoreData._id, storeParams, dispatch, state)
      .then((success) => {
        if (success) {
          setFormOpen(false);
        }
      })
  }

  const handleFormOpen = () => {
    setFormOpen(!formOpen);
  };
  useEffect(() => {
    console.log("form loaded")
  }, []);

  useEffect(() => {
    if (!formOpen) {
      window.scrollTo({ top: 0, behavior: "smooth" });
    }
  }, [formOpen])
  useEffect(() => {
    if (title && description) {
      setNewForm(false);
    }
  }, [title, description]);


  return (
    <div id="storeFormHolder">
      {
        !newForm ?
          <React.Fragment>
            <Grid.Row>
              <Grid.Column mobile={16} tablet={14} computer={14}>
                <h1>Details</h1>
                <div className="storeFormHolderDetails">
                  <h3>Store title:</h3>
                  <p>{currentStoreData.title}</p>
                </div>
                <div className="storeFormHolderDetails">
                  <h3>Store description:</h3>
                  <p>{currentStoreData.description}</p>
                </div>
                <div className="storeFormHolderTimestamps">
                  <span>Created At: {ConvertDate.international(currentStoreData.createdAt)}</span>
                  <span>Edited At: {ConvertDate.international(currentStoreData.editedAt)}</span>
                </div>
              </Grid.Column>
            </Grid.Row>
            <Grid.Row>
              <Grid.Column mobile={16} tablet={14} computer={14}>
                <StoreImgPreviewHolder state={state} dispatch={dispatch} />
              </Grid.Column>
            </Grid.Row>
            <StoreImageUplForm />
          </React.Fragment>
          : null
      }
      <Grid.Row>
        <Grid.Column mobile={16} tablet={15} computer={14}>
          <Button  id="storeFormToggleBtn" onClick={handleFormOpen} content={ !formOpen ? "Open Form" : "Close Form"}></Button>
          {
            formOpen ? <StoreForm 
                        title={title} 
                        description={description} 
                        handleCreateStore={handleCreateStore}
                        handleUpdateStore={handleUpdateStore}
                      /> 
                      : null
          }
         
        </Grid.Column>
      </Grid.Row>
    </div>
  );
}

export default StoreFormHolder;

