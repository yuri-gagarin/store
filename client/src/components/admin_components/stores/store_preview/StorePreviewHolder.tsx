import React from "react";
import { IGlobalAppState, AppAction } from "../../../../state/Store";
// additional components //
import StorePreview from "./StorePreview";
import LoadingScreen from "../../miscelaneous/LoadingScreen";

interface Props {
  state: IGlobalAppState;
  dispatch: React.Dispatch<AppAction>
}

const StorePreviewHolder: React.FC<Props> = ({ state, dispatch }): JSX.Element => {
  const { loadedStores, loading } = state.storeState;
  if (loading) {
    return (
      <LoadingScreen />
    );
  } else {
    return (
      <div>
        {
          loadedStores.map((store) => {
            return (
              <StorePreview />
            )
          })
        }
      </div>
    );
  }
};

export default StorePreviewHolder;