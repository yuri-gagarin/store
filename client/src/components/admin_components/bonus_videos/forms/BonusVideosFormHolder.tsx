import React, { useState, useEffect, useContext } from "react";
import { Button, Grid } from "semantic-ui-react";
// css imports //
import "./css/bonusVideoFormHolder.css";
// additional components //
import BonusVideoForm from "./BonusVideoForm";
import BonusVideoPreview from "../bonus_videos_preview/BonusVideoPreview";
// state //
import { Store } from "../../../../state/Store";
// api and ui actions //
import { createBonusVideo, editBonusVideo } from "../actions/APIBonusVideoActions";
import { closeBonusVideoForm, openBonusVideoForm } from "../actions/UIBonusVideoActions";
// helpers //
import { ConvertDate } from "../../../helpers/displayHelpers";
import { checkSetValues } from "../../../helpers/validationHelpers";
// type definitions //
import { BonusVideoData, FormState } from "../type_definitions/bonusVideoTypes"
import { RouteComponentProps } from "react-router-dom";

interface Props { }

const BonusVideoFormHolder: React.FC<Props> = ({ }): JSX.Element => {
  const { state, dispatch} = useContext(Store);
  // data for component //
  const { currentBonusVideoData, bonusVideoFormOpen } = state.bonusVideoState;
  const { description, youTubeURL, vimeoURL, createdAt, editedAt } = currentBonusVideoData;

  // form toggle listener //
  const toggleBonusVidform = () => {
    bonusVideoFormOpen ? closeBonusVideoForm(dispatch) : openBonusVideoForm(dispatch);
  }
  // CREATE - EDIT methods //
  const handleCreateBonusVideo = ({ description, youTubeURL, vimeoURL }: FormState): void => {
    const bonusVideoData: BonusVideoData = {
      description,
      youTubeURL,
      vimeoURL
    };

    createBonusVideo(bonusVideoData, dispatch)
      .then((_) => {
        closeBonusVideoForm(dispatch);
      })
      .catch((_) => {
        // handle error show modal ? //
      });
  };
  const handleUpdateBonusVideo = ({ description, youTubeURL, vimeoURL }: FormState): void => {
    const bonusVideoParams: BonusVideoData = {
      description,
      youTubeURL,
      vimeoURL
    };

    editBonusVideo(currentBonusVideoData._id, bonusVideoParams, dispatch, state)
      .then((_) => {
          closeBonusVideoForm(dispatch)
      })
      .catch((_) => {
        // handle an error or show a modal ? //
      })
  };
  // lifecycle hooks //
  useEffect(() => {
    if (!bonusVideoFormOpen) {
      window.scrollTo({ top: 0, behavior: "smooth" });
    }
  }, [ bonusVideoFormOpen ]);
  // component render //
  return (
    <div id="bonusVideoFormHolder">
      {
        checkSetValues<IBonusVideoData>(currentBonusVideoData) ?
          <React.Fragment>
            <Grid.Row id="bonusVideoFormDetails">
              <Grid.Column mobile={16} tablet={14} computer={14}>
                <h1>Details</h1>
                <div className="bonusVideoFormHolderYouTubeUrl">
                  <h3>Video youTube URL:</h3>
                  <p>{youTubeURL}</p>
                </div>
                <div className="bonusVideoFormHolderVimeoUrl">
                  <h3>Video vimeo URL:</h3>
                  <p>{vimeoURL}</p>
                </div>
                <div className="bonusVideoFormHolderDescription">
                  <h3>Video description:</h3>
                  <p>{description}</p>
                </div>
                <div className="bonusVideoFormHolderTimestamps">
                  <span>Created At: <strong>{ConvertDate.international(createdAt)}</strong></span>
                  <span>Edited At: <strong>{ConvertDate.international(editedAt)}</strong></span>
                </div>
                <div className="bonusVideoFormPreview">
                  <BonusVideoPreview bonusVideo={currentBonusVideoData} />
                </div>
              </Grid.Column>
            </Grid.Row>
          </React.Fragment>
          : null
      }
      <Grid.Row>
        <Grid.Column mobile={16} tablet={15} computer={14}>
          <Button  
            id="bonusVideoFormToggleBtn" 
            onClick={toggleBonusVidform} 
            content={ bonusVideoFormOpen ? "Close Form" : "Open Form"}
          />
          {
            bonusVideoFormOpen ? 
            <BonusVideoForm 
              description={description} 
              youTubeURL={youTubeURL}
              vimeoURL={vimeoURL}
              handleCreateBonusVideo={handleCreateBonusVideo}
              handleUpdateBonusVideo={handleUpdateBonusVideo}
            /> 
            : null
          }
         
        </Grid.Column>
      </Grid.Row>
    </div>
  );
};

export default BonusVideoFormHolder;

