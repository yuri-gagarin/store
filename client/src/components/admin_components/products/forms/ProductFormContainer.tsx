import React, { useState, useEffect, useContext } from "react";
import { Button, Grid } from "semantic-ui-react";
// client routing //
import { RouteComponentProps, withRouter } from "react-router-dom";
// css imports //
import "./css/productFormContainer.css";
// additional components //
import ProductForm from "./ProductForm";
import ProductImgPreviewContainer from "../image_preview/ProductImgPreviewContainer";
import LoadingBar from "../../miscelaneous/LoadingBar";
import ProductImgUplForm from "./ProductImgUplForm";
// state //
import { Store } from "../../../../state/Store";
// api and ui actions //
import { createProduct, editProduct } from "../actions/APIProductActions";
import { closeProductForm, openProductForm, clearProductAPIError } from "../actions/UIProductActions";
// helpers //
import { ConvertDate } from "../../../helpers/displayHelpers";
import { checkSetValues } from "../../../helpers/validationHelpers";
// types 
import { ProductData, ProductFormState } from "../type_definitions/productTypes";
import ErrorBar from "../../miscelaneous/ErrorBar";

interface ProductFormContState {
  productFormState: ProductFormState;
  newForm: boolean;
}

interface Props extends RouteComponentProps {
 
}

const ProductFormContainer: React.FC<Props> = ({ history }): JSX.Element => {
  const { state, dispatch } = useContext(Store);
  const { loading, currentProductData, productFormOpen, error } = state.productState;
  const { name, description, details, price, createdAt, editedAt } = currentProductData;
  // local component state //
  // const [ formOpen, setFormOpen ] = useState<boolean>(false);
  const [ productFormContState, setProductFormContState ] = useState<ProductFormContState>({
    newForm: true,
    productFormState: { name: "", price: "", details: "", description: ""}
  });

  // ProductForm toggle //
  const toggleProductForm = () => {
    productFormOpen ? closeProductForm(dispatch) : openProductForm(dispatch);
  };
  // CLEAR Product API ERRROR //
  const clearCurrentError = () => {
    clearProductAPIError(dispatch);
  };
  // API call handlers CREATE - EDIT //
  const handleCreateProduct = (productFormData: ProductFormState): void => {
    const productData: ProductData = {
      ...productFormData,
      images: currentProductData.images
    };
    createProduct(productData, dispatch)
      .then((_) => {
        // product created //
        closeProductForm(dispatch);
      })
      .catch((_) => {
        // handle error or show modal //
        // set form state //
        // console.log("create error occured")
        setProductFormContState({
          ...productFormContState,
          productFormState: { ...productData }
        });
      });
  };
  const handleUpdateProduct = (productFormData: ProductFormState): void => {
    const productParams: ProductData = {
      ...productFormData,
      images: currentProductData.images
    };
    editProduct(currentProductData._id, productParams, dispatch, state)
      .then((_) => {
        closeProductForm(dispatch)
      })
      .catch((_) => {
        // handle error show modal //
      });
  };
  // lifecycle hooks //
  useEffect(() => {
    if (!productFormOpen) {
      window.scrollTo({ top: 0, behavior: "smooth" });
    }
  }, [ productFormOpen ]);
  useEffect(() => {
    if (checkSetValues(currentProductData)) {
      const { name, price, description, details } = currentProductData;
      setProductFormContState({
        newForm: false,
        productFormState: {
          name,
          price,
          description,
          details
        }
      });
    }
  }, [ currentProductData ]);
  
  // component return //
  return (
    <div id="productFormContainer">
      { loading ? <LoadingBar /> : null }
      {
        !productFormContState.newForm ?
          <div id={"productFormContainerDetails"}>
            <Grid.Row> 
              <Grid.Column mobile={16} tablet={14} computer={14}>
                <h1>Details</h1>
                <div className="productFormContainerDetailsItem">
                  <h3>Product name:</h3>
                  <p>{name}</p>
                </div>
                <div className="productFormContainerDetailsItem">
                  <h3>Product price:</h3>
                  <p>{price}</p>
                </div>
                <div className="productFormContainerDetailsItem">
                  <h3>Product description:</h3>
                  <p>{description}</p>
                </div>
                <div className="productFormContainerDetailsItem">
                  <h3>Product details:</h3>
                  <p>{details}</p>
                </div>
                <div className="productFormContainerTimestamps">
                  <span>Created At: <strong>{ConvertDate.international(createdAt)}</strong></span>
                  <span>Edited At: <strong>{ConvertDate.international(editedAt)}</strong></span>
                </div>
              </Grid.Column>
            </Grid.Row>
            <Grid.Row>
              <Grid.Column mobile={16} tablet={14} computer={14}>
                <ProductImgPreviewContainer state={state} dispatch={dispatch} />
              </Grid.Column>
            </Grid.Row>
            <ProductImgUplForm />
          </div>
          : null
      }
      <Grid.Row>
        <Grid.Column mobile={16} tablet={15} computer={14}>
          <Button  
            id="productFormToggleBtn"
            onClick={toggleProductForm} 
            content={ !productFormOpen ? "Open Form" : "Close Form"}
          />
          { error ? <ErrorBar clearError={clearCurrentError} error={error} /> : null }
          {
            productFormOpen ? 
              <ProductForm 
                newForm={productFormContState.newForm}
                handleCreateProduct={handleCreateProduct}
                handleUpdateProduct={handleUpdateProduct}
                { ...productFormContState.productFormState }
              /> 
            : null
          }
         
        </Grid.Column>
      </Grid.Row>
    </div>
  );
};
// export without router for tests //
export { ProductFormContainer };
// default export //
export default withRouter(ProductFormContainer);

