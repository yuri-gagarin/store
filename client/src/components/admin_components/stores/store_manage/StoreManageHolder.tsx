import React, { useContext, useEffect } from "react";
import { Button, Grid } from "semantic-ui-react";
import StoreFormHolder from "../forms/StoreFormHolder";
import StoreCard from "./StoreCard";
import { withRouter, RouteComponentProps, useRouteMatch, Route } from "react-router-dom";
import { Store } from "../../../../state/Store";
import { getAllStores } from "../actions/APIstoreActions";

interface Props extends RouteComponentProps {};

const StoreManageHolder: React.FC<Props> = ({ history }): JSX.Element => {
  const { state, dispatch } = useContext(Store);
  const { loadedStores } = state.storeState;
  const match = useRouteMatch("/admin/home/my_store/manage");

  const handleBack = () => {
    history.goBack();
  }
  useEffect(() => {
    getAllStores(dispatch);
  }, []); 

  return (
    <Grid padded>
       
      <Route path={match?.url + "/edit"}> 
        <Grid.Row>
          <Grid.Column>
            <h3>Editing Store: { state.storeState.currentStoreData.title }</h3>
            <Button inverted color="green" content="Back" onClick={handleBack}></Button>
          </Grid.Column>
        </Grid.Row>
        <StoreFormHolder state={state} dispatch={dispatch} />
      </Route>
      <Route exact path={match?.url}>
        {
          loadedStores.map((store) => {
            return (
              <StoreCard 
                key={store._id}
                _id={store._id}
                title={store.title}
                description={store.description}
                imageCount={store.images.length}
                createdAt={store.createdAt}
                editedAt={store.editedAt}
                state={state}
                dispatch={dispatch}
              />
            );
          })
        }
      </Route>
     
    </Grid>
  );
};

export default withRouter(StoreManageHolder);