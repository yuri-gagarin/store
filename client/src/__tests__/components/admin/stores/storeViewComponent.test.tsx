import React, { ReactElement } from "react";
import { Menu } from "semantic-ui-react";
// react router //
import  { BrowserRouter as Router, Switch } from "react-router-dom";
// testing imports //
import { mount, shallow, ReactWrapper , ShallowWrapper} from "enzyme";
// component imports //
import StoreViewComponent from "../../../../components/admin_components/stores/StoreView";
import AdminStoreMenu from "../../../../components/admin_components/menus/AdminStoreMenu";
import StorePreviewHolder from "../../../../components/admin_components/stores/store_preview/StorePreviewHolder";
import { StoreFormHolder } from "../../../../components/admin_components/stores/forms/StoreFormHolder";
import { StoreManageHolder } from "../../../../components/admin_components/stores/store_manage/StoreManageHolder";
// additional dependencies //

describe("StoreView Component test", () => {
  // TEST StoreViewComponent render //
  describe("StoreView Component render test", () => {
    let component: ShallowWrapper;
    beforeAll(() => {
      component = shallow(
        <StoreViewComponent />
      );
    });
    it("Should Properly render StoreView", () => {
      expect(component).toMatchSnapshot();
    });
    it("Should render conditional routes", () => {
      expect(component.find("Switch")).toHaveLength(1);
    });
    it("Should have Render three children route componets", () => {
      const switchComponent = component.find(Switch);
      expect(switchComponent.props().children).toHaveLength(3);
    });
    it("Should have proper client routes and conditionally render correct Components", () => {
      const switchComponent = component.find(Switch);
      type RouteMap = {
        [key: string]: string
      }
      let map: RouteMap = {};

      const routeComponents = switchComponent.props().children as ReactElement[];
      for (const component of routeComponents) {
        if (component.props.children[1].type.WrappedComponent) {
          // to check if a component is wrapped in a WithRouter() function //
          map[component.props.path as string] = component.props.children[1].type.WrappedComponent;

        } else {
          map[component.props.path as string] = component.props.children[1].type;
        }
      }
      expect(map["/admin/home/my_stores/all"]).toBe(StorePreviewHolder);
      expect(map["/admin/home/my_stores/create"]).toBe(StoreFormHolder);
      expect(map["/admin/home/my_stores/manage"]).toBe(StoreManageHolder);
    });
  });
  // END StoreViewComponent render tests //
  // TEST AdminStoreMenu within component and Link button tests //
  describe("StoreView Component button actions", () => {
    let component: ReactWrapper;

    beforeAll(() => {
      component = mount(
        <Router>
          <StoreViewComponent />
        </Router>
      );
    });

    describe("AdminStoreMenu", () => {

      it("Should render Admin Store Menu", () => {
        console.log(component.find("withRouter").length)
        expect(component.find("AdminStoreMenu")).toHaveLength(1);
      });
      it("Should have 3 main navigation links", () => {
        const wrapper = component.find(AdminStoreMenu);
        const links = wrapper.find(Menu.Item)
        expect(links.length).toEqual(3);
      });
      it("'View All Stores' link should properly function", () => {
        window.scrollTo = jest.fn();
        const viewAllLink = component.find(AdminStoreMenu).find(Menu.Item).at(0);
        viewAllLink.simulate("click");
        component.update();
        expect(component.find(StorePreviewHolder).length).toEqual(1);
        expect(component.find(StoreFormHolder).length).toEqual(0);
        expect(component.find(StoreManageHolder).length).toEqual(0);
      });
      it("'Create Store' link should properly function", () => {
        window.scrollTo = jest.fn();
        const viewAllLink = component.find(AdminStoreMenu).find(Menu.Item).at(1);
        viewAllLink.simulate("click");
        component.update();
        expect(component.find(StorePreviewHolder).length).toEqual(0);
        expect(component.find(StoreFormHolder).length).toEqual(1);
        expect(component.find(StoreManageHolder).length).toEqual(0);
      });
      it("'Manage Stores' link should properly function", () => {
        window.scrollTo = jest.fn();
        const viewAllLink = component.find(AdminStoreMenu).find(Menu.Item).at(2);
        viewAllLink.simulate("click");
        component.update();
        expect(component.find(StorePreviewHolder).length).toEqual(0);
        expect(component.find(StoreFormHolder).length).toEqual(0);
        expect(component.find(StoreManageHolder).length).toEqual(1);
      });

    });
  });
  // END AdminStoreMenu button tests //

});