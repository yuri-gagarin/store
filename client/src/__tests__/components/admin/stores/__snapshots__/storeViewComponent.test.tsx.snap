// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoreView Component render tests StoreView Component default render test Should Properly render StoreView 1`] = `
<div
  id="adminStoreViewHolder"
>
  <withRouter(AdminStoreMenu)
    dispatch={[Function]}
  />
  <Switch>
    <Route
      path="/admin/home/my_stores/all"
    >
      <Component
        height="100px"
        width="100%"
      />
      <withRouter(StorePreviewHolder) />
    </Route>
    <Route
      path="/admin/home/my_stores/create"
    >
      <Component
        height="100px"
        width="100%"
      />
      <withRouter(StoreFormHolder) />
    </Route>
    <Route
      path="/admin/home/my_stores/manage"
    >
      <Component
        height="100px"
        width="100%"
      />
      <withRouter(StoreManageHolder) />
    </Route>
  </Switch>
</div>
`;

exports[`StoreView Component test StoreView Component render test Should Properly render StoreView 1`] = `
<div
  id="adminStoreViewHolder"
>
  <withRouter(AdminStoreMenu)
    dispatch={[Function]}
  />
  <Switch>
    <Route
      path="/admin/home/my_stores/all"
    >
      <Component
        height="100px"
        width="100%"
      />
      <StorePreviewHolder
        dispatch={[Function]}
        state={
          Object {
            "bonusVideoState": Object {
              "currentBonusVideoData": Object {
                "_id": "",
                "createdAt": "",
                "description": "",
                "vimeoURL": "",
                "youTubeURL": "",
              },
              "error": null,
              "loadedBonusVideos": Array [],
              "loading": false,
              "responseMsg": "",
            },
            "productState": Object {
              "currentProductData": Object {
                "_id": "",
                "createdAt": "",
                "description": "",
                "details": "",
                "images": Array [],
                "name": "",
                "price": "",
              },
              "error": null,
              "loadedProducts": Array [],
              "loading": false,
              "productFormOpen": false,
              "responseMsg": "",
            },
            "serviceState": Object {
              "currentServiceData": Object {
                "_id": "",
                "createdAt": "",
                "description": "",
                "images": Array [],
                "name": "",
                "price": "",
              },
              "error": null,
              "loadedServices": Array [],
              "loading": false,
              "responseMsg": "",
              "serviceFormOpen": false,
            },
            "storeItemState": Object {
              "currentStoreItemData": Object {
                "_id": "",
                "categories": Array [],
                "createdAt": "",
                "description": "",
                "details": "",
                "images": Array [],
                "name": "",
                "price": "",
                "storeId": "",
                "storeName": "",
              },
              "error": null,
              "loadedStoreItems": Array [],
              "loading": false,
              "numberOfItems": 0,
              "responseMsg": "",
              "storeItemFormOpen": false,
            },
            "storeState": Object {
              "currentStoreData": Object {
                "_id": "",
                "createdAt": "",
                "description": "",
                "images": Array [],
                "title": "",
              },
              "error": null,
              "loadedStores": Array [],
              "loading": false,
              "responseMsg": "",
              "storeFormOpen": false,
            },
          }
        }
      />
    </Route>
    <Route
      path="/admin/home/my_stores/create"
    >
      <Component
        height="100px"
        width="100%"
      />
      <withRouter(StoreFormHolder) />
    </Route>
    <Route
      path="/admin/home/my_stores/manage"
    >
      <Component
        height="100px"
        width="100%"
      />
      <withRouter(StoreManageHolder) />
    </Route>
  </Switch>
</div>
`;
